openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -768,6 +768,87 @@ def _get_encrypted_request_headers(request):
         return encrypt_for_log(str(request.headers), getattr(settings, 'SAFE_SESSIONS_DEBUG_PUBLIC_KEY', None))
 
 
+class EmailChangeSessionInvalidationMiddleware(MiddlewareMixin):
+    """
+    Middleware responsible for invalidating user sessions
+    by detecting a mismatch between the user's email in
+    the session and request.user.email.
+
+    This middleware ensures that the sessions in other browsers
+    are invalidated when a user changes their email in one browser.
+    The active session in which the email change is made will remain valid.
+
+    The user's email is stored in their session during login
+    and gets updated when the user changes their email.
+    This middleware checks for any mismatch between the stored email
+    and the current user's email in each request, and if found,
+    it invalidates/flushes the session and mark cookies for deletion in request
+    which are then deleted in the process_response of SafeSessionMiddleware.
+    """
+
+    def process_request(self, request):
+        """
+        Invalidate the user session if there's a mismatch
+        between the email in the user's session and request.user.email.
+        """
+        if request.user.is_authenticated:
+            user_session_email = request.session.get('email', None)
+            are_emails_mismatched = user_session_email is not None and request.user.email != user_session_email
+            EmailChangeSessionInvalidationMiddleware._set_session_email_match_custom_attribute(are_emails_mismatched)
+            if settings.ENFORCE_SESSION_EMAIL_MATCH and are_emails_mismatched:
+                # Flush the session and mark cookies for deletion.
+                request.session.flush()
+                request.user = AnonymousUser()
+                _mark_cookie_for_deletion(request)
+
+    def process_response(self, request, response):
+        """
+        Update the logged-in cookies if the email change was marked
+        """
+        if (
+            settings.ENFORCE_SESSION_EMAIL_MATCH
+            and request.user.is_authenticated
+            and self._is_email_change_marked(request)
+        ):
+            # Update the JWT cookies with new user email
+            response = set_logged_in_cookies(request, response, request.user)
```

Comment:
I've got a number of thoughts related to this. I was confused, and it took me a bit to clarify thing for my self.
* [confirmation requested] Is this a bug fix? In the past, did we simply not update the cookies to match the email change when it was first changed?
* [request] I think we should rename `EmailChangeSessionInvalidationMiddleware` to `EmailChangeMiddleware`, and document that it has two jobs: 1) Upon email change, it will update the session's email and update the JWT cookie to match the new email, and 2) Will invalidate any future session where the user's email does not match its session email.
* [nit] Since `email_changed` (the name you are using for the request flag) could be misinterpreted as a mismatch (e.g. change detected), maybe update this to `email_change_requested` to better clarify that it is only used during a session in which the email change was being made?
* [nit] Lastly, I don't think this needs to be protected by `ENFORCE_SESSION_EMAIL_MATCH`, but I'll leave that up to you. Worst case this would be extra processing if we had a bug, but this isn't as dangerous and clearing the session.

Reply:
Thankyou for the thoughts and suggestions.

> [confirmation requested] Is this a bug fix? In the past, did we simply not update the cookies to match the email change when it was first changed?

We had the cookie update logic in the `confirm_email_change` function which is now moved to `process_response` function of middleware as per the suggestion in this [comment](https://github.com/openedx/edx-platform/pull/33846#discussion_r1433236315). Plus the custom `update_jwt_cookies_in_response` function was removed and we used `set_logged_in_cookies` instead to avoid possible sync issues. 

This was to make sure that JWT authentication also fails on email change.

Please let me know if this is what you were asking?

Reply:
Yes, you are right. We were not updating the JWT cookie previously. So yes, this fixes that too.